### PR TITLE
[CALCITE-3789] Support validation of UNNEST multiple array columns like Presto

### PR DESCRIPTION
**Pull request by my7ym**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java [hunk 1]
```diff
@@ -97,6 +97,17 @@ public SqlUnnestOperator(boolean withOrdinality) {
     return builder.build();
   }
 
+  private boolean allowFlattenStruct(SqlOperatorBinding operatorBinding) {
+    if (!(operatorBinding instanceof SqlCallBinding)) {
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java]: `allowFlattenStruct` -> `allowAliasUnnestArrayColumns` ?

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java]: Another class type for operator binding is `RexCallBinding`, but why we return true for that ?

File: core/src/main/java/org/apache/calcite/sql/validate/DelegatingScope.java [hunk 2]
```diff
@@ -298,7 +298,11 @@ public SqlQualified fullyQualify(SqlIdentifier identifier) {
       for (; i > 0; i--) {
         final SqlIdentifier prefix = identifier.getComponent(0, i);
         resolved.clear();
-        resolve(prefix.names, nameMatcher, false, resolved);
+        resolve(
+            prefix.names,
+            nameMatcher,
+            validator.getConformance().allowAliasUnnestArrayColumns(),
+            resolved);
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/validate/DelegatingScope.java]: The code may works for your case, but you have changed a common sql identifier resolving logic with a special conformance, which seems not right way to go ~

The conformance switch logic should always happen in SqlValidator, we can pass along some switch flags into the scope if necessary.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java [hunk 3]
```diff
@@ -73,7 +73,9 @@
 
   /** Conformance value that instructs Calcite to use SQL semantics
    * consistent with Microsoft SQL Server version 2008. */
-  SQL_SERVER_2008;
+  SQL_SERVER_2008,
+
+  PRESTO;
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java]: PRESTO is not really a db, it behaves more like a computation engine, i'm not sure if we should add a new conformance for it.

File: core/src/main/java/org/apache/calcite/sql/validate/DelegatingScope.java (hunk 2 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql/validate/DelegatingScope.java]: Sorry a little confused here. Here I am directly using conformance as the flag. I feel like it's more concise than an explicit flag like 

SqlValidatorImpl.java::
boolean allowAliasUnnestArrays = validator.getConformance().allowAliasUnnestArrayColumns;
DelegatingScope.setAllowAliasUnnestArrays(allowAliasUnnestArrays);

DeletatingScope.java::
Just replace validator.getConformance().allowAliasUnnestArrayColumns with the explicit flag.

And it seems using conformance in scope happens before. e.g. OrderByScope.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java (hunk 3 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java]: That's right. I am also debating on this. I could delete it or rename it. But I think the conformance is only for SQL string validation, so whether or not DB is kinda fine. But it's your call.

File: core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java (hunk 1 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java]: @danny0405 Took a second thought on the function name, the result of the function is actually whether **disallow flatten struct in UNNEST**. So maybe disallowFlattenStructInUnnest? Let me know. Thanks.

**Comment by my7ym** [comment]
@danny0405 Just a kindly reminder, could you please take another look when available? Thanks a lot!

**Comment by my7ym** [comment]
@danny0405 Could you help take another look at this PR? Thanks!

**Comment by danny0405** [comment]
I'm planning to review again this weekend, very busy with the release work of Calcite 1.22 and also my own compony work, sorry for that, Just like Julian said, add a new SqlConformance is really a big change, for "big" i mean, we should keep all the changes correct.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java (hunk 3 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java]: Presto has its own SQL parser & validator & type system, so it counts as a 'dialect' (conformance) for these purposes.

Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java]: I'll add further comments to https://issues.apache.org/jira/browse/CALCITE-3789.

**Comment by my7ym** [comment]
> I'm planning to review again this weekend, very busy with the release work of Calcite 1.22 and also my own company work, sorry for that, Just like Julian said, add a new SqlConformance is really a big change, for "big" i mean, we should keep all the changes correct.

Make sense. Just send a reminder and get a rough timeline. Thanks @danny0405

**Comment by my7ym** [comment]
@danny0405 @julianhyde Just a soft reminder on this PR. No rush. Thanks!

**Comment by danny0405** [comment]
Reviewing now, may take some time ~

File: core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java [hunk 4]
```diff
@@ -451,12 +449,9 @@ public boolean addRule(RelOptRule rule) {
...
 
     final String ruleName = rule.toString();
-    if (ruleNames.put(ruleName, rule.getClass())) {
-      Set<Class> x = ruleNames.get(ruleName);
-      if (x.size() > 1) {
-        throw new RuntimeException("Rule description '" + ruleName
-            + "' is not unique; classes: " + x);
-      }
+    if (ruleNames.put(ruleName, rule) != null) {
+      throw new RuntimeException("Rule description '" + ruleName
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java]: Did you merge the master code instead of rebase ? The change seems not that necessary.

File: core/src/main/java/org/apache/calcite/sql/SqlDialectFactoryImpl.java [hunk 5]
```diff
@@ -144,7 +145,8 @@ public SqlDialect create(DatabaseMetaData databaseMetaData) {
     } else if (databaseProductName.startsWith("HP Neoview")) {
       return new NeoviewSqlDialect(c);
     } else if (upperProductName.contains("POSTGRE")) {
-      return new PostgresqlSqlDialect(c);
+      return new PostgresqlSqlDialect(
+          c.withDataTypeSystem(PostgresqlSqlDialect.POSTGRESQL_TYPE_SYSTEM));
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/SqlDialectFactoryImpl.java]: Also a unnecessary change.

**Comment by danny0405** [comment]
Thanks @my7ym , could you rebase you code to remove the unnecessary change ?

File: core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java (hunk 4 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java]: @danny0405 Done

File: core/src/main/java/org/apache/calcite/sql/SqlDialectFactoryImpl.java (hunk 5 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql/SqlDialectFactoryImpl.java]: Done

File: core/src/main/java/org/apache/calcite/sql/validate/DelegatingScope.java (hunk 2 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/validate/DelegatingScope.java]: The current fix seems not right, the nested field can be seen if we set up the record type to `StructKind.PEEK_FIELDS`, for the test cases you gave:
```java
sql("select e.ENAME\n"
        + "from dept_nested as d CROSS JOIN\n"
        + " UNNEST(d.employees) as t(e)")
      .withConformance(SqlConformanceEnum.PRESTO).columnType("VARCHAR(10) NOT NULL");
```
There `t(e)` table has record type with `e` as field, the `e` can be seen and if you change the nested employee type with  code snippet:
```java
final RelDataType empRecordType = typeFactory.builder()
      .add("EMPNO", intType)
      .add("ENAME", varchar10Type)
      .add("DETAIL", typeFactory.builder()
          .add("SKILLS", array(skillRecordType)).build())
      .kind(StructKind.PEEK_FIELDS) // here is the line you can tweak
      .build();
```

So, in total, this line change is not necessary, the field access should not be controlled by the `SqlConformance`.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlAbstractConformance.java [hunk 6]
```diff
@@ -111,4 +111,7 @@ public boolean allowQualifyingCommonColumn() {
     return SqlConformanceEnum.DEFAULT.allowQualifyingCommonColumn();
   }
 
+  public boolean allowAliasUnnestArrayColumns() {
+    return SqlConformanceEnum.DEFAULT.allowAliasUnnestArrayColumns();
+  }
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlAbstractConformance.java]: How about the `MultiSet` type.

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java [hunk 7]
```diff
@@ -7540,6 +7540,42 @@ public void _testGroupExpressionEquivalenceParams() {
...
+      .withExtendedCatalog()
+      .withCaseSensitive(false)
+      .withConformance(SqlConformanceEnum.PRESTO).ok();
+    sql("select e.ENAME\n"
+        + "from dept_nested as d CROSS JOIN\n"
+        + " UNNEST(d.employees) as t(e)")
+      .withConformance(SqlConformanceEnum.PRESTO).columnType("VARCHAR(10) NOT NULL");
+    sql("select d.deptno, e.detail, k.empno\n"
+      + "from dept_nested as d CROSS JOIN\n"
+      + " UNNEST(d.employees, d.employees) as t(e, k)")
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java]: You should add a new nested `StructKind.PEEK_FIELDS` type and write test with that.

File: core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java (hunk 1 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java]: I mean `RexCallBinding` comes from building the call from `RelBuilder`, but why we return true for that, confusing for me.

**Comment by danny0405** [comment]
We should also add a test case for at least one SqlToRelConverterTest.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlAbstractConformance.java (hunk 6 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlAbstractConformance.java]: Could you elaborate here?

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java (hunk 7 above)
Comment (my7ym) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java]: Will do

**Comment by my7ym** [comment]
> We should also add a test case for at least one SqlToRelConverterTest.

Will do it in the future if needed. I am still unclear that how this should be represented as RelNode. (discussed in https://issues.apache.org/jira/browse/CALCITE-3787)

 Is it OK for you?

File: core/src/main/java/org/apache/calcite/sql/validate/SqlAbstractConformance.java (hunk 6 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlAbstractConformance.java]: I mean you `unnest` support `multiset` type which is also supported in your code. But the method name here does not represent that.

Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlAbstractConformance.java]: Got it. Do you have a better name in mind? I picked array because it seems to be a little more common (at least for me). But I am OK with other names. Let me know.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlAbstractConformance.java]: Just `allowAliasUnnestColumns` seems better, no need to specify the operand type, i think.

File: core/src/main/java/org/apache/calcite/sql/validate/DelegatingScope.java (hunk 2 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql/validate/DelegatingScope.java]: make sense. Will do

File: core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java (hunk 1 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java]: Again, this is original behavior. I feel like for a PR, it could be better to only contain all the relevant changes without changing existing behaviors for irrelevant flows. 

There are 8 other subclasses of `SqlOperatorBinding` and I guess several of them should not be here. If you believe this is a serious issue and should be addressed, we should have another refactor for it.

Let me know if you insist on gating the `RexCallBinding`, I will add another commit to do it.

File: core/src/test/java/org/apache/calcite/test/catalog/MockCatalogReaderSimple.java [hunk 8]
```diff
@@ -155,6 +155,8 @@ public MockCatalogReaderSimple(RelDataTypeFactory typeFactory,
     deptNestedTable.addColumn("NAME", fixture.varchar10Type);
     deptNestedTable.addColumn("SKILL", fixture.skillRecordType);
     deptNestedTable.addColumn("EMPLOYEES", fixture.empListType);
+    deptNestedTable.addColumn("ADMINS", fixture.varchar5ArrayType);
+    deptNestedTable.addColumn("OFFICES", fixture.rectilinearPeekCoordMultisetType);
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/test/catalog/MockCatalogReaderSimple.java]: Do not modify existing table/schema to avoid introducing unnecessary change of plans.

File: core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java (hunk 1 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java]: I know you only to support `SqlCallBinding` for unnest alias columns, but the logic seems weird, can you change
`allowFlattenStruct` => `allowAliasUnnestColumns` ?

File: core/src/test/java/org/apache/calcite/test/catalog/MockCatalogReaderSimple.java (hunk 8 above)
Comment (my7ym) [reply core/src/test/java/org/apache/calcite/test/catalog/MockCatalogReaderSimple.java]: I am really confused about your review.

What do you mean by 

> You should add a new nested StructKind.PEEK_FIELDS type and write test with that.

Right now you again think we should create a new table/schema to avoid it. I guess it's expected for you to know that this table is used in a lot of places. Maybe I am wrong here?

And what's the drawbacks of so? It's test fixtures and should be shared with multiple tests. Do we need to create a new test fixtures for every PR if it changes other test assertions?

If you insist, please think it again and let me know if I create another table with something like
`  deptNestedExtended.addColumn("NAME", fixture.varchar10Type);
    deptNestedExtended.addColumn("ADMINS", fixture.varchar5ArrayType);
    deptNestedExtended.addColumn("EMPLOYEES", fixture.empListType);
    deptNestedExtended.addColumn("OFFICES", fixture.rectilinearPeekCoordMultisetType);
`

File: core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java (hunk 1 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java]: I think it looks even more confusing. If you read the code carefully enough, it is **NOT** allowAliasUnnestColumns. If you prefer the name change, I think `disallowAliasUnnestColumn` is more appropriate.

File: core/src/test/java/org/apache/calcite/test/catalog/MockCatalogReaderSimple.java (hunk 8 above)
Comment (danny0405) [reply core/src/test/java/org/apache/calcite/test/catalog/MockCatalogReaderSimple.java]: -1 because we should never break existing test because of not relevant change.

I don’t care if you reuse the test code base but at least let’s not break other tests.

**Comment by danny0405** [comment]
> > We should also add a test case for at least one SqlToRelConverterTest.
> 
> Will do it in the future if needed. I am still unclear that how this should be represented as RelNode. (discussed in https://issues.apache.org/jira/browse/CALCITE-3787)
> 
> Is it OK for you?
> 
> EDIT: Some other reasons:
> 
> 1. We still plan to communicate with Presto using SQL.
> 2. RelNode -> SqlNode is not quite mature I believe, but I may be wrong there.
> 3. The optimization enabled by RelNode is helpful. But not a high-pri since Presto will also optimize the internal execution plan.

Usually we have 2 kinds of use cases for Calcite:

- Generates RelNode/RexNode to adapter other downstream planners
- Generates RelNode/RexNode and unparse it as special SQL dialect to interpretation through JDBC.

Either way we should generates a working plan, that is why I suggest to add SqlToRel test. For more general, we should add operator test and SQL ITCase if we change an operator semantics. But for you case, I think a SqlToRelConverter test is okey.

File: core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java (hunk 1 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java]: Couldn’t you use !allowAliasUnnestColumns ?

Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java]: That's a good idea.

Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java]: Good idea. Will do

File: core/src/test/java/org/apache/calcite/test/catalog/MockCatalogReaderSimple.java (hunk 8 above)
Comment (my7ym) [reply core/src/test/java/org/apache/calcite/test/catalog/MockCatalogReaderSimple.java]: OK. I will add a new table and add above columns in it and use that for tests.

**Comment by my7ym** [comment]
> > > We should also add a test case for at least one SqlToRelConverterTest.
> > 
> > 
> > Will do it in the future if needed. I am still unclear that how this should be represented as RelNode. (discussed in https://issues.apache.org/jira/browse/CALCITE-3787)
> > Is it OK for you?
> > EDIT: Some other reasons:
> > 
> > 1. We still plan to communicate with Presto using SQL.
> > 2. RelNode -> SqlNode is not quite mature I believe, but I may be wrong there.
> > 3. The optimization enabled by RelNode is helpful. But not a high-pri since Presto will also optimize the internal execution plan.
> 
> Usually we have 2 kinds of use cases for Calcite:
> 
> * Generates RelNode/RexNode to adapter other downstream planners
> * Generates RelNode/RexNode and unparse it as special SQL dialect to interpretation through JDBC.
> 
> Either way we should generates a working plan, that is why I suggest to add SqlToRel test. For more general, we should add operator test and SQL ITCase if we change an operator semantics. But for you case, I think a SqlToRelConverter test is okey.

OK. I could do that for something like unnest(employees) as t(e), but I am not sure what's a valid plan for unnest(employees, admins) as t(e, a). Can I just add one test case for something like unnest(employees) as t(e)? Thanks.

**Comment by my7ym** [comment]
> > > We should also add a test case for at least one SqlToRelConverterTest.
> > 
> > 
> > Will do it in the future if needed. I am still unclear that how this should be represented as RelNode. (discussed in https://issues.apache.org/jira/browse/CALCITE-3787)
> > Is it OK for you?
> > EDIT: Some other reasons:
> > 
> > 1. We still plan to communicate with Presto using SQL.
> > 2. RelNode -> SqlNode is not quite mature I believe, but I may be wrong there.
> > 3. The optimization enabled by RelNode is helpful. But not a high-pri since Presto will also optimize the internal execution plan.
> 
> Usually we have 2 kinds of use cases for Calcite:
> 
> * Generates RelNode/RexNode to adapter other downstream planners
> * Generates RelNode/RexNode and unparse it as special SQL dialect to interpretation through JDBC.
> 
> Either way we should generates a working plan, that is why I suggest to add SqlToRel test. For more general, we should add operator test and SQL ITCase if we change an operator semantics. But for you case, I think a SqlToRelConverter test is okey.

I have added tests for it. But it needs changes on SqlToRelConverter. 

The change is because usually for STRUCT, Uncollect will flatten the type but for this feature, it should not. This is also aligned with the registered type during validation.

I am not sure whether this should be the best way to implement it. Let me know. Thanks.

File: core/src/test/java/org/apache/calcite/test/catalog/MockCatalogReaderSimple.java (hunk 8 above)
Comment (my7ym) [reply core/src/test/java/org/apache/calcite/test/catalog/MockCatalogReaderSimple.java]: Done

**Comment by my7ym** [comment]
> We should also add a test case for at least one SqlToRelConverterTest.

Done

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java [hunk 9]
```diff
@@ -1304,7 +1304,31 @@ public final Sql sql(String sql) {
...
+  /**
+   * Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-3789">[CALCITE-3789]
+   * Support validation of UNNEST multiple array columns like Presto</a>.
+   */
+  @Test public void testAliasUnnestArrayPlanWithSingleColumn() {
+    final String sql = "select d.deptno, employee.empno\n"
+        + "from dept_nested_expanded as d,\n"
+        + " UNNEST(d.employees) as t(employee)";
+    sql(sql).conformance(SqlConformanceEnum.PRESTO).decorrelate(false).ok();
```
Comment (my7ym) [inline core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java]: The reason of turning off `decorrelate` is because during flattening, the row type of Uncollect is not fully flattened, but the index is calculated on a fully flattened base (RelStructuredTypeFlattener.postFlattenSize). So the type mismatches validated query. Not sure whether it will be a big issue tho.

File: core/src/main/java/org/apache/calcite/rel/core/Uncollect.java [hunk 10]
```diff
@@ -130,25 +139,32 @@ public static RelDataType deriveUncollectRowType(RelNode rel,
...
           .build();
     }
 
+
+    int aliasIndex = 0;
     for (RelDataTypeField field : fields) {
       if (field.getType() instanceof MapSqlType) {
         builder.add(SqlUnnestOperator.MAP_KEY_COLUMN_NAME, field.getType().getKeyType());
         builder.add(SqlUnnestOperator.MAP_VALUE_COLUMN_NAME, field.getType().getValueType());
       } else {
```
Comment (my7ym) [inline core/src/main/java/org/apache/calcite/rel/core/Uncollect.java]: I am also not sure what will be the best way to handle MAP here.

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java (hunk 9 above)
Comment (danny0405) [reply core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java:1319]: It is, you should fix the `RelStructuredTypeFlattener`.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 11]
```diff
@@ -2047,8 +2047,15 @@ protected void convertFrom(
...
       call = (SqlCall) from;
-      convertFrom(bb, call.operand(0));
-      if (call.operandCount() > 2
+      SqlNode firstOperand = call.operand(0);
+
+      if (firstOperand.getKind() == SqlKind.UNNEST) {
+        convertUnnest(bb, (SqlCall) firstOperand, getFieldAliases(call));
+      } else {
+        convertFrom(bb, firstOperand);
+      }
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2056]: The operator row type belongs to the Validator namespace, i think we should fix it there instead of `SqlToRelConverter`.

Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2056]: Yea, I considered it. For example, for query
{code}
select d.deptno, e, k.empno\n"
        + "from dept_nested_expanded as d CROSS JOIN\n"
        + " UNNEST(d.admins, d.employees) as t(e, k)
{code}
After validation, it will become
{code}
select d.deptno, t.e, t.k.empno\n"
        + "from dept_nested_expanded as d CROSS JOIN\n"
        + " UNNEST(d.admins, d.employees) as t(e, k)
{code}
The failure comes from the Project. Since the rowtype of Uncollect is already flattened, so Calcite could not find the path t.e and t.k.empo from the input, which is the Uncollect RelNode.

So my change will keep them nested, as in SqlUnnestOperator, and I think it's also aligned with the expected behavior, which is unnesting array but not unbox the component struct. 

Could you elaborate on your thoughts on AliasNamespace? And if possible, could you provide an ideal logical plan looks like? Thanks!

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2056]: For type flattern, fix the `RelStructuredTypeFlattener` for the operator you care about instead of the `SqlToRelConverter`.

Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2056]: @danny0405 I don't quite your point. If you are talking about the decorrelate, I will fix it. But this part of change is not about decorrelate. Could you elaborate more on how does your fix work in more detail? If it's a fundamental change, could you please post it in the ticket too? I will also describe my plan there. Thanks.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java [hunk 12]
```diff
@@ -431,7 +438,22 @@
...
    * true otherwise.
    */
   boolean allowQualifyingCommonColumn();
+
+  /**
+   * Whether directly alias array columns in UNNEST.
+   *
+   * <p>E.g. in UNNEST(a_array, b_array) AS T(a, b),
+   * a and b will be aliases of a_array and b_array
+   * respectively
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java:451]: What will be the behavior of that expression if `allowAliasUnnestColumns` is false?

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java [hunk 13]
```diff
@@ -73,7 +73,11 @@
 
   /** Conformance value that instructs Calcite to use SQL semantics
    * consistent with Microsoft SQL Server version 2008. */
-  SQL_SERVER_2008;
+  SQL_SERVER_2008,
+
+  /** Conformance value that instructs Calcite to use SQL semantics
+   * consistent with Presto. */
+  PRESTO;
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java:80]: alphabetical

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java [hunk 14]
```diff
@@ -330,10 +341,19 @@ public boolean allowExtendedTrim() {
     case STRICT_92:
     case STRICT_99:
     case STRICT_2003:
+    case PRESTO:
       return false;
     default:
       return true;
     }
   }
+  public boolean allowAliasUnnestColumns() {
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java:350]: need space before this function, no space after.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java (hunk 12 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java:451]: What is the most logical place for this function inside SqlConformance? I bet it's not at the very end.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java (hunk 11 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2056]: I agree, AliasNamespace sounds like a good direction to explore.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 15]
```diff
@@ -2219,6 +2207,42 @@ protected void convertFrom(
     }
   }
 
+  private List<String> getFieldAliases(SqlCall asCall) {
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2210]: This method needs javadoc

Is returning null the best thing? If it's illegal to call the method, you should throw. I hate nullable returns.

Conformance usually applies during validation, not during sql-to-rel. You know validation has succeeded at this point. If only one thing is possible, you should just do it.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 16]
```diff
@@ -2219,6 +2207,42 @@ protected void convertFrom(
...
+    List<String> columnAliases = new ArrayList<>();
+    for (SqlNode sqlNode : Util.skip(asCall.getOperandList(), 2)) {
+      columnAliases.add(((SqlIdentifier) sqlNode).getSimple());
+    }
+    return columnAliases;
+  }
+
+  private void convertUnnest(Blackboard bb, SqlCall from, List<String> fieldAliases) {
+    SqlCall call;
+    call = from;
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2224]: assign on one line and make final? or eliminate a variable?

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 17]
```diff
@@ -2219,6 +2207,42 @@ protected void convertFrom(
...
+    final List<SqlNode> nodes = call.getOperandList();
+    final SqlUnnestOperator operator = (SqlUnnestOperator) call.getOperator();
+    for (SqlNode node : nodes) {
+      replaceSubQueries(bb, node, RelOptUtil.Logic.TRUE_FALSE_UNKNOWN);
+    }
+    final List<RexNode> exprs = new ArrayList<>();
+    final List<String> fieldNames = new ArrayList<>();
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2231]: you don't need two separate arrays. `RelBuilder.project` will apply aliases if you have used `RelBuilder.alias` to wrap expressions in 'as'. Use `project` rather than `projectNamed`.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 18]
```diff
@@ -2219,6 +2207,42 @@ protected void convertFrom(
...
+    for (Ord<SqlNode> node : Ord.zip(nodes)) {
+      exprs.add(bb.convertExpression(node.e));
+      fieldNames.add(validator.deriveAlias(node.e, node.i));
+    }
+    RelNode child =
+        (null != bb.root) ? bb.root : LogicalValues.createOneRow(cluster);
+    relBuilder.push(child).projectNamed(exprs, fieldNames, false);
+
+    Uncollect uncollect =
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2240]: it's about time we added `RelBuilder.uncollect`

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java [hunk 19]
```diff
@@ -7540,6 +7540,47 @@ public void _testGroupExpressionEquivalenceParams() {
         .fails("Column 'C1' not found in any table");
   }
 
+  /**
```
Comment (julianhyde) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:7543]: We need some negative tests, to prove that UNNEST is illegal in non-Presto conformance.

Consider splitting this method up, and adding comments. It's not clear what each test case is testing.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java (hunk 11 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2075]: @danny0405 @julianhyde I revisited the flow. I will use UNNEST(d.employees) as t(k) as an example.

In validation phase, since this is a AS call and Calcite produces an AliasNamespace. And since we disable flatten of STRUCT type for SqlUnnestOperator, the output type is "expected", with one field `k` and its type is employee struct type.

During SqlToRel, the problem is in bb.lookupExp in convertIdentifer (e.g. T.K.EMPO). Basically the identifier could be resolved from namespace, the type is also expected. When `lookup` RexNode in inputs, the type output is wrong because the input type is already flattened by Uncollect:deriveRowType. 

Basically, we either modify Uncollect.deriveRowType to produce the properly aliased unflattend row type, or we modify the `lookup` Rexnode logic, and maybe some other places. The second one is pretty de-centralized so I suggest to take the first one.

For the first solution, my implementation is one option. Another option is to directly use the right namespace's rowType in Uncollect.deriveRowType. One question is that I see generally RelNode type deriving is separate from validation results (namespace's rowType). May I know why? And do you think it's gonna work for this use case? 

Thanks.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2075]: Did you notice that i have given a fix patch on the issue ? [1]

[1] https://issues.apache.org/jira/browse/CALCITE-3789

Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2075]: @danny0405 Sorry was a little busy recently and just noticed it. Added one nit commit. Your way to plug in the aliases is really clean 👍

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java (hunk 15 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2229]: This method will be deleted in @danny0405 patch. So I will resolve this comment.

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java (hunk 9 above)
Comment (my7ym) [reply core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java:1319]: Done in @danny0405 patch. Will resolve this comment.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java (hunk 11 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2075]: @danny0405 patch greatly improves the first approach. Resolve this comment.

**Comment by danny0405** [comment]
If you think my fix is okey, can you check it in this patch, thanks ~

**Comment by my7ym** [comment]
> If you think my fix is okey, can you check it in this patch, thanks ~

Left another comment. I thought you preferred to keep them separate so I did not merge them. Will do after that comment is resolved! Thanks!

**Comment by danny0405** [comment]
I have fired a new fix at https://github.com/danny0405/calcite/commit/b98d0cd04e2e9c66393a255bd6bcddf4e8e8fe24, please check if you have time.

**Comment by my7ym** [comment]
> I have fired a new fix at https://github.com/danny0405/calcite/tree/fix-3789, please check if you have time.

👍 Will do and let you know.

**Comment by danny0405** [comment]
> > I have fired a new fix at https://github.com/danny0405/calcite/tree/fix-3789, please check if you have time.
> 
> 👍 Will do and let you know.
> 
> EIDT: Just two nit comments. Solid fix. Let me know if you want to update it or I could directly resolve them when I merge your fix. Thanks!

Directly resolve them when you merge the fix, thanks for the contribution ~

**Comment by my7ym** [comment]
> > > I have fired a new fix at https://github.com/danny0405/calcite/tree/fix-3789, please check if you have time.
> > 
> > 
> > 👍 Will do and let you know.
> > EIDT: Just two nit comments. Solid fix. Let me know if you want to update it or I could directly resolve them when I merge your fix. Thanks!
> 
> Directly resolve them when you merge the fix, thanks for the contribution ~

Done. Not sure whether I should make the commit message & JIRA title the same. Feel free to modify either of them. Thanks!
